File: packages/domain/src/__tests__/ids.test.ts
import { describe, expect, it } from "vitest";
import { AttachmentId, NoteId, NotebookId, TagId, UserId } from "../ids.js";

const VALID_UUID = "550e8400-e29b-41d4-a716-446655440000";

describe("branded ID constructors", () => {
  it.each([
    ["UserId", UserId],
    ["NotebookId", NotebookId],
    ["TagId", TagId],
    ["NoteId", NoteId],
    ["AttachmentId", AttachmentId],
  ] as const)("%s accepts a valid UUID and returns the same raw string", (_, ctor) => {
    expect(ctor(VALID_UUID)).toBe(VALID_UUID);
  });

  it("rejects empty strings", () => {
    expect(() => UserId("")).toThrow(/cannot be empty/);
  });

  it("rejects non-UUID strings", () => {
    expect(() => NoteId("not-a-uuid")).toThrow(/must be a UUID/);
  });

  it("rejects UUIDs with the wrong version nibble", () => {
    // Version nibble is the first hex of the third group; '6' is invalid (v1-5 only).
    expect(() => NoteId("550e8400-e29b-61d4-a716-446655440000")).toThrow(/must be a UUID/);
  });
});

File: packages/android/core-domain/src/test/kotlin/app/notes/domain/IdsTest.kt
package app.notes.domain

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class IdsTest {
    private val validUuid = "550e8400-e29b-41d4-a716-446655440000"

    @Test
    fun `accepts a valid UUID and stores the raw string`() {
        val id = NoteId(validUuid)
        assertEquals(validUuid, id.raw)
    }

    @Test
    fun `rejects empty strings`() {
        val ex = assertThrows(IllegalArgumentException::class.java) { UserId("") }
        assert(ex.message!!.contains("UserId"))
    }

    @Test
    fun `rejects non-UUID strings`() {
        assertThrows(IllegalArgumentException::class.java) { NoteId("not-a-uuid") }
    }

    @Test
    fun `rejects UUIDs with the wrong version nibble`() {
        assertThrows(IllegalArgumentException::class.java) {
            NoteId("550e8400-e29b-61d4-a716-446655440000")
        }
    }

    @Test
    fun `value classes are zero-cost at runtime but distinct at compile time`() {
        val noteId = NoteId(validUuid)
        val userId = UserId(validUuid)
        // Same underlying raw string, but the type system keeps them apart.
        assertEquals(noteId.raw, userId.raw)
    }
}
